### PR TITLE
Morello freebsd64 support

### PR DESCRIPTION
**Pull request by zxombie**
Add support for `COMPAT_FREEBSD64` on Morello. While here also handle user visible registers.

### CODE REVIEW

File: sys/arm64/include/ucontext.h [hunk 1]
```diff
@@ -71,18 +71,43 @@ struct __mcontext {
 	struct fpregs	mc_fpregs;
 	int		mc_flags;
 #define	_MC_FP_VALID	0x1		/* Set when mc_fpregs has valid data */
+#define	_MC_CAP_VALID	(1u<<31)	/* Set when mc_capregs has valid data */
```
Comment (zxombie) [inline sys/arm64/include/ucontext.h:74]: I've changed this from morello-dev so it won't conflict if upstream adds flags, e.g. for SVE.

File: sys/arm64/include/ucontext.h [hunk 2]
```diff
@@ -71,18 +71,43 @@ struct __mcontext {
...
 #if (defined(_KERNEL) && __has_feature(capabilities)) || \
     defined(__CHERI_PURE_CAPABILITY__)
 	__uint32_t	mc_spsr;
 	__uint64_t	mc_spare[8];	/* Space for expansion, set to zero */
 #else
 	int		mc_pad;		/* Padding */
-	__uint64_t	mc_spare[8];	/* Space for expansion, set to zero */
+	__uint64_t	mc_spare[7];	/* Space for expansion, set to zero */
+	__uint64_t	mc_capregs;
```
Comment (zxombie) [inline sys/arm64/include/ucontext.h:82]: I've moved this to the end so it won't conflict if we use the spare values upstream.

File: sys/arm64/arm64/freebsd64_machdep.c [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+	.sv_szsigcode	= &freebsd64_szsigcode,
+	.sv_name	= "FreeBSD ELF64",
+	.sv_coredump	= __elfN(coredump),
+	.sv_imgact_try	= NULL,
+	.sv_minsigstksz	= MINSIGSTKSZ,
+	.sv_minuser	= VM_MIN_ADDRESS,
+	.sv_maxuser	= VM_MAXUSER_ADDRESS,
+	.sv_usrstack	= USRSTACK,
+	.sv_psstrings	= FREEBSD64_PS_STRINGS,
+	.sv_stackprot	= VM_PROT_READ | VM_PROT_WRITE,
```
Comment (jrtc27) [inline sys/arm64/arm64/freebsd64_machdep.c:94]: This needs to be RW_CAP if you want it to work until you pull in John's PTE commit, I believe?

File: sys/arm64/arm64/freebsd64_machdep.c [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+{
+	const uintcap_t *creg;
+	register_t *greg;
+	u_int i;
+
+	memset(mc64, 0, sizeof(*mc64));
+	creg = (uintcap_t *)&mc->mc_capregs;
+	greg = (register_t *)&mc64->mc_gpregs;
+	for (i = 0; i < CONTEXT64_COPYREGS; i++)
+		greg[i] = (__cheri_addr register_t)creg[i];
```
Comment (jrtc27) [inline sys/arm64/arm64/freebsd64_machdep.c:147]: Shouldn't need the `__cheri_addr` for uintcap_t->integer casts these days.

File: sys/arm64/arm64/freebsd64_machdep.c [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+	mc64->mc_gpregs.gp_elr =
+	    (__cheri_offset register_t)mc->mc_capregs.cap_elr;
```
Comment (jrtc27) [inline sys/arm64/arm64/freebsd64_machdep.c:149]: Hm, cheri_getoffset might be better these days.

File: sys/arm64/arm64/freebsd64_machdep.c [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+		    sizeof(mc.mc_capregs)), &mc.mc_capregs,
+		    sizeof(mc.mc_capregs));
+		if (error)
+			return (error);
+
+		/* XXX: Permit userland to change GPRs for sigreturn? */
+
+		/* Honor 64-bit PC. */
+		mc.mc_capregs.cap_elr = (uintcap_t)cheri_setoffset(
+		    (void * __capability)mc.mc_capregs.cap_elr,
```
Comment (jrtc27) [inline sys/arm64/arm64/freebsd64_machdep.c:191]: Casts shouldn't be necessary any more.

File: sys/arm64/arm64/freebsd64_machdep.c [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+		    mcp->mc_gpregs.gp_elr);
+	} else {
+		creg = (uintcap_t *)&mc.mc_capregs;
+		greg = (register_t *)&mcp->mc_gpregs;
+		for (i = 0; i < CONTEXT64_COPYREGS; i++)
+			creg[i] = (uintcap_t)greg[i];
+
+		mc.mc_capregs.cap_elr = (uintcap_t)cheri_setoffset(
+		    (void * __capability)td->td_frame->tf_elr,
```
Comment (jrtc27) [inline sys/arm64/arm64/freebsd64_machdep.c:200]: Ditto.

File: sys/arm64/arm64/freebsd64_machdep.c [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+
+	/* Copy the sigframe out to the user's stack. */
+	if (copyoutcap(&frame, __USER_CAP_OBJ(fp), sizeof(*fp)) != 0) {
+		/* Process has trashed its stack. Kill it. */
+		CTR2(KTR_SIG, "sendsig: sigexit td=%p fp=%p", td, fp);
+		PROC_LOCK(p);
+		sigexit(td, SIGILL);
+	}
+
+	tf->tf_x[0]= sig;
```
Comment (jrtc27) [inline sys/arm64/arm64/freebsd64_machdep.c:295]:
```suggestion
	tf->tf_x[0] = sig;
```

File: sys/arm64/arm64/machdep.c [hunk 9]
```diff
@@ -1410,6 +1448,14 @@ DB_SHOW_COMMAND(specialregs, db_show_spregs)
 #define	PRINT_REG(reg)	\
     db_printf(__STRING(reg) " = %#016lx\n", READ_SPECIALREG(reg))
 
+#if __has_feature(capabilities)
+#define	PRINT_REG_CAP(reg)						\
+do {									\
+    void * __capability _tmp = (void * __capability)READ_SPECIALREG_CAP(reg); \
+    db_printf(__STRING(reg) " = %#lp\n", _tmp);				\
```
Comment (jrtc27) [inline sys/arm64/arm64/machdep.c:1455]:
```suggestion
    db_printf(__STRING(reg) " = %#.16lp\n", _tmp);				\
```
to match the integer registers

File: sys/arm64/include/armreg.h [hunk 10]
```diff
@@ -64,6 +64,16 @@
 #define	WRITE_SPECIALREG(reg, _val)					\
 	__asm __volatile("msr	" __STRING(reg) ", %0" : : "r"((uint64_t)_val))
 
+#if __has_feature(capabilities)
+#define	READ_SPECIALREG_CAP(reg)						\
+({	uintcap_t _val;							\
+	__asm __volatile("mrs	%0, " __STRING(reg) : "=&C" (_val));	\
```
Comment (jrtc27) [inline sys/arm64/include/armreg.h:70]: Doesn't need to be volatile but I guess it's consistent with the non-_CAP version upstream

File: sys/arm64/include/ucontext.h [hunk 11]
```diff
@@ -71,18 +71,43 @@ struct __mcontext {
...
 typedef struct __mcontext mcontext_t;
 
+#ifdef COMPAT_FREEBSD64
+#include <compat/freebsd64/freebsd64_signal.h>
+typedef struct __mcontext64 {
+	struct gpregs	mc_gpregs;
+	struct fpregs	mc_fpregs;
+	int		mc_flags;
+#define	_MC_FP_VALID	0x1		/* Set when mc_fpregs has valid data */
+#define	_MC_CAP_VALID	(1u<<31)	/* Set when mc_capregs has valid data */
```
Comment (jrtc27) [inline sys/arm64/include/ucontext.h:95]: Already defined

**Comment by jrtc27** [comment]
Please make all your commit messages say that they're for Morello (or arm64) in the subject; I gave them all a "morello:" prefix in the previous PR which makes it nice and consistent in the logs, but other approaches are fine too, just not something vague that doesn't make it clear what architecture it applies to.

**Comment by brooksdavis** [comment]
I've gone and addressed all but one of @jrtc27's comments. Hopefully I didn't step on any toes in the progress.

**Comment by jrtc27** [comment]
Result of `interdiff <(git diff HEAD...321f6ce) <(git diff HEAD...22f810c)` (one day GitHub will display _that_ instead):
```diff
diff -u b/sys/arm64/arm64/freebsd64_machdep.c b/sys/arm64/arm64/freebsd64_machdep.c
--- b/sys/arm64/arm64/freebsd64_machdep.c
+++ b/sys/arm64/arm64/freebsd64_machdep.c
@@ -91,7 +91,7 @@
 	.sv_maxuser	= VM_MAXUSER_ADDRESS,
 	.sv_usrstack	= USRSTACK,
 	.sv_psstrings	= FREEBSD64_PS_STRINGS,
-	.sv_stackprot	= VM_PROT_READ | VM_PROT_WRITE,
+	.sv_stackprot	= VM_PROT_RW_CAP,
 	.sv_copyout_auxargs = __elfN(freebsd_copyout_auxargs),
 	.sv_copyout_strings = freebsd64_copyout_strings,
 	.sv_setregs	= exec_setregs,
@@ -144,9 +144,8 @@
 	creg = (uintcap_t *)&mc->mc_capregs;
 	greg = (register_t *)&mc64->mc_gpregs;
 	for (i = 0; i < CONTEXT64_COPYREGS; i++)
-		greg[i] = (__cheri_addr register_t)creg[i];
-	mc64->mc_gpregs.gp_elr =
-	    (__cheri_offset register_t)mc->mc_capregs.cap_elr;
+		greg[i] = (register_t)creg[i];
+	mc64->mc_gpregs.gp_elr = cheri_getoffset(mc->mc_capregs.cap_elr);
 	mc64->mc_gpregs.gp_spsr = mc->mc_spsr;
 	mc64->mc_flags = mc->mc_flags;
 	if (mc->mc_flags & _MC_FP_VALID)
@@ -187,8 +186,7 @@
 		/* XXX: Permit userland to change GPRs for sigreturn? */
 
 		/* Honor 64-bit PC. */
-		mc.mc_capregs.cap_elr = (uintcap_t)cheri_setoffset(
-		    (void * __capability)mc.mc_capregs.cap_elr,
+		mc.mc_capregs.cap_elr = cheri_setoffset(mc.mc_capregs.cap_elr,
 		    mcp->mc_gpregs.gp_elr);
 	} else {
 		creg = (uintcap_t *)&mc.mc_capregs;
@@ -292,7 +290,7 @@
 		sigexit(td, SIGILL);
 	}
 
-	tf->tf_x[0]= sig;
+	tf->tf_x[0] = sig;
 	tf->tf_x[1] = (uintcap_t)&fp->sf_si;
 	tf->tf_x[2] = (uintcap_t)&fp->sf_uc;
 
diff -u b/sys/arm64/arm64/machdep.c b/sys/arm64/arm64/machdep.c
--- b/sys/arm64/arm64/machdep.c
+++ b/sys/arm64/arm64/machdep.c
@@ -1452,7 +1452,7 @@
 #define	PRINT_REG_CAP(reg)						\
 do {									\
     void * __capability _tmp = (void * __capability)READ_SPECIALREG_CAP(reg); \
-    db_printf(__STRING(reg) " = %#lp\n", _tmp);				\
+    db_printf(__STRING(reg) " = %#.16lp\n", _tmp);				\
 } while (0)
 #endif
 
diff -u b/sys/arm64/include/ucontext.h b/sys/arm64/include/ucontext.h
--- b/sys/arm64/include/ucontext.h
+++ b/sys/arm64/include/ucontext.h
@@ -91,8 +91,6 @@
 	struct gpregs	mc_gpregs;
 	struct fpregs	mc_fpregs;
 	int		mc_flags;
-#define	_MC_FP_VALID	0x1		/* Set when mc_fpregs has valid data */
-#define	_MC_CAP_VALID	(1u<<31)	/* Set when mc_capregs has valid data */
 	int		mc_pad;		/* Padding */
 	__uint64_t	mc_spare[7];	/* Space for expansion, set to zero */
 	__uint64_t	mc_capregs;
```

**Comment by brooksdavis** [comment]
> I think you folded the fixup into the wrong commit?

Arg, I caught that error once and then messed it up the next time though.

File: sys/arm64/arm64/freebsd64_machdep.c (hunk 7 above)
Comment (jrtc27) [reply sys/arm64/arm64/freebsd64_machdep.c:198]: Still outstanding

**Review by jrtc27 — APPROVED** [review]
(assuming it builds in CI)

**Pull request merged**